packages/graphiql/src/utility/QueryStore.js

```diff
@@ -64,13 +65,38 @@
   }
 
   push(item) {
-    this.items.push(item);
-    this.save();
-  }
+    const items = [...this.items, item];
+
+    if (this.maxSize && items.length > this.maxSize) {
+      items.shift();
+    }
+
+    let done = false;
+    let success = false;
+    let tries = 0;
+    while (!done) {
+      tries++;
+      // do not use this.save because we only want to add to the list there's no storage error
+      const response = this.storage.set(this.key, JSON.stringify({ [this.key]: items }));;
 
-  shift() {
-    this.items.shift();
-    this.save();
+      if (response.error === null) {
+        done = true;
+        success = true;
+      } else if (response.quotaError && this.maxSize) {
+        items.shift();
+      } else {
+        done = true;
+      }
+
+      // Never try more than 5 times
+      if (tries > 5) {
+        done = true;
+      }
+    }
+
+    if (success) {
+      this.items = items;
+    }
   }
 
   save() {
```

Comment:
Can we delete the `save` method now or is it still used someplace else?

Reply:
`save` is still used by the `delete` and `edit` functions since we don't check for quota on these.